src/extensions/surveys.tsx

```diff
@@ -187,6 +193,84 @@ export function generateSurveys(posthog: PostHog) {
     }, 3000)
 }
 
+export function usePopupVisibility(
+    survey: Survey,
+    posthog: PostHog | undefined,
+    delay: number,
+    isPreviewMode: boolean
+) {
+    const [isPopupVisible, setIsPopupVisible] = useState(isPreviewMode || delay === 0)
+    const [isSurveySent, setIsSurveySent] = useState(false)
+
+    useEffect(() => {
+        if (isPreviewMode || !posthog) {
+            return
+        }
+
+        const handleSurveyClosed = () => {
+            setIsPopupVisible(false)
+        }
+
+        const handleSurveySent = () => {
+            if (!survey.appearance?.displayThankYouMessage) {
+                return setIsPopupVisible(false)
+            }
+
+            setIsSurveySent(true)
+
+            if (survey.appearance?.autoDisappear) {
+                setTimeout(() => {
+                    setIsPopupVisible(false)
+                }, 5000)
+            }
+        }
+
+        window.addEventListener('PHSurveyClosed', handleSurveyClosed)
+        window.addEventListener('PHSurveySent', handleSurveySent)
+
+        if (delay > 0) {
+            const timeoutId = setTimeout(() => {
+                setIsPopupVisible(true)
+
+                window.dispatchEvent(new Event('PHSurveyShown'))
```

Comment:
can we abstract this into a local function call that does all of these things?

Reply:
sorry, must've missed this!  Just addressed it.